aptos-move/aptos-aggregator/src/aggregator_change_set.rs

```diff
@@ -2,331 +2,447 @@
+                AggregatorValue::Snapshot(delta.apply_to(base_value.into_aggregator_value()?)?)
             },
+            SnapshotDerived { formula, .. } => {
+                AggregatorValue::Derived(formula.apply(base_value.into_snapshot_value()?))
+            },
+        })
+    }
+}
+
+impl AggregatorChange {
+    // When squashing a new change on top of the old one, sometimes we need to know the change
+    // from a different AggregatorID to be able to merge them together.
+    // In particular SnapshotDelta represents a change from the aggregator at the beginning of the transaction,
+    // and squashing changes where the aggregator will be at the beginning of the transaction.
+    // For example, let’s say we have two change sets that we need to squash:
+    // change1: agg1 -> Delta(+3)
+    // change2: agg1 -> Delta(+6), snap1 -> (base=agg1, Delta(+2))
+    // the correct squashing of snapshot depends on the change for the base aggregator. I.e. the correct output would be:
+    // agg1 -> Delta(+9), snap(base=agg1, Delta(+5))
+    pub fn get_merge_dependent_id(&self) -> Option<AggregatorID> {
+        use AggregatorApplyChange::*;
+        use AggregatorChange::*;
+
+        match self {
+            // Only SnapshotDelta merging logic depends on current aggregator change
+            Apply(SnapshotDelta {
+                base_aggregator, ..
+            }) => Some(*base_aggregator),
+            Create(_) | Apply(AggregatorDelta { .. } | SnapshotDerived { .. }) => None,
         }
     }
 
-    /// Applies next aggregator change on top of self, merging two changes together. This is a reverse
-    /// of `merge_with_previous_aggregator_change`.
-    pub fn merge_with_next_aggregator_change(
-        &mut self,
+    /// Applies next AggregatorChange on top of the previous state.
+    /// prev_change is AggregatorChange for the same AggregatorID
+    /// prev_dependent_change is AggregatorChange for the get_merge_dependent_id()
+    pub fn merge_two_changes(
+        prev_change: Option<&AggregatorChange>,
+        prev_dependent_change: Option<&AggregatorChange>,
         next_change: &AggregatorChange,
-    ) -> PartialVMResult<()> {
-        // Now self follows the other delta.
-        let mut prev_change = next_change.clone();
-        std::mem::swap(self, &mut prev_change);
-
-        // Perform the merge.
-        self.merge_with_previous_aggregator_change(&prev_change)?;
-        Ok(())
+    ) -> PartialVMResult<AggregatorChange> {
+        use AggregatorApplyChange::*;
+        use AggregatorChange::*;
+        use AggregatorValue::*;
+
+        // There are only few valid cases for merging:
+        // - next_change being AggregatorDelta, and prev_change being Aggregator Create or Delta
+        // - next_change being SnapshotDelta, and prev_dependent_change being Aggregator Create or Delta
+        // everything else is invalid for various reasons
+        match (&prev_change, &prev_dependent_change, next_change) {
```

Comment:
Merging 3 changes in "merge_two_changes" function is confusing.
And I didn't understand why we need to merge 3 changes at a time. In the above example,
change1: agg1 -> Delta(+3)
change2: agg1 -> Delta(+6), snap1 -> (base=agg1, Delta(+2))
We can merge [ change1(agg1), change2(agg1) ] and merge [ change1(agg1), change2(snap1) ] separately right?

Reply:
will do